staging/src/k8s.io/component-base/zpages/flagz/flagz.go

```diff
@@ -0,0 +1,126 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package flagz
+
+import (
+	"bytes"
+	"fmt"
+	"math/rand"
+	"net/http"
+	"sync"
+	"text/template"
+
+	"k8s.io/klog/v2"
+)
+
+const (
+	flagzTemplate = `
+{{.ComponentName}} flags
+
+Warning: This endpoint is not meant to be machine parseable, has no formatting compatibility guarantees and is for debugging purposes only.
+
+`
+)
+
+var (
+	flagzRegistry   = &registry{}
+	flagzSeparators = []string{":", ": ", "=", " "}
+)
+
+type registry struct {
+	response bytes.Buffer
+	once     sync.Once
+}
+
+type mux interface {
+	Handle(path string, handler http.Handler)
+}
+
+type templateData struct {
+	ComponentName string
+}
+
+func Install(m mux, componentName string, flagReader Reader) {
+	flagzRegistry.installHandler(m, componentName, flagReader)
+}
+
+func (reg *registry) installHandler(m mux, componentName string, flagReader Reader) {
+	tmpl, err := initializeTemplates()
+	if err != nil {
+		klog.Errorf("error while parsing gotemplate: %v", err)
+		return
+	}
+	m.Handle("/flagz", reg.handleFlags(componentName, tmpl, flagReader))
+}
+
+func (reg *registry) handleFlags(componentName string, tmpl *template.Template, flagReader Reader) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
```

Comment:
make sure this errors if a request that does not accept text/plain comes in... in the future, if / when we add structured output, we don't want a future client that requests *only* application/json to get a successful response from an older server returning text/plain

see goautoneg.ParseAccept, iterating over parsed accept types, ensuring one has Type = "text" or "*" and SubType = "plain" or "*" and Params are empty or only have keys/values that match "charset"/"utf-8" (case-insensitive)

Reply:
sounds good, added validation for request's Accept header.

Reply:
> Params are empty or only have keys/values that match "charset"/"utf-8" (case-insensitive)

looks like that bit is still outstanding

Reply:
Sorry about that, messed up the rebase I guess. Added now. thanks for the catch!